include/wlr/types/wlr_input_method.h

```diff
@@ -0,0 +1,74 @@
+#ifndef WLR_TYPES_WLR_INPUT_METHOD_H
+#define WLR_TYPES_WLR_INPUT_METHOD_H
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+#include <wlr/interfaces/wlr_input_device.h>
+#include <wlr/interfaces/wlr_keyboard.h>
+
+struct wlr_input_panel {
+	struct wl_global *panel;
+	struct wl_list surfaces; // struct wlr_input_panel_surface*
+	struct {
+		struct wl_signal new_surface; // struct wlr_input_panel_surface*
+	} events;
+};
+
+struct wlr_input_panel_surface {
+	struct wlr_surface *surface;
+	struct wlr_input_panel *panel;
+
+	struct wl_list link;
+
+	struct {
+		struct wl_signal destroy; // struct wlr_input_panel_surface*
+	} events;
+};
+
+struct wlr_input_method {
+	struct wl_global *input_method;
+	struct wl_resource *resource;
+	struct {
+		struct wl_signal new_context; // struct wlr_input_method_context*
+		// TODO: create roots_keyboard (seat_add_keyboard) on this; modify roots_keyboard to support backend-less
```

Comment:
I think this should be more along the lines of wlr_input_method creating a wlr_keyboard which roots_keyboard should just be able to pick up without changes. Maybe some silly changes to automatically associate it with the relevant seat if `wlr_keyboard_is_input_method` returns true.